clients/adapter/src/main/java/org/eclipse/hono/adapter/client/command/CommandRouterClient.java

```diff
@@ -0,0 +1,87 @@
+     * for the <em>gatewayId</em> parameter.
+     *
+     * @param tenantId The tenant id.
+     * @param deviceId The device id.
+     * @param gatewayId The gateway id (or the device id if the last message came from the device directly).
+     * @param context The currently active OpenTracing span context or {@code null} if no span is currently active.
+     *            An implementation should use this as the parent for any span it creates for tracing
+     *            the execution of this operation.
+     * @return A future indicating whether the operation succeeded or not.
+     * @throws NullPointerException if device id or gateway id is {@code null}.
+     */
+    Future<Void> setLastKnownGatewayForDevice(String tenantId, String deviceId, String gatewayId, SpanContext context);
+
+    /**
+     * Registers a protocol adapter instance as the consumer of command & control messages
+     * for a device.
+     *
+     * @param tenantId The tenant id.
+     * @param deviceId The device id.
+     * @param adapterInstanceId The protocol adapter instance id.
+     * @param lifespan The lifespan of the registration entry. Using a negative duration or {@code null} here is
+     *                 interpreted as an unlimited lifespan. Only the number of seconds in the given duration
+     *                 will be taken into account.
+     * @param context The currently active OpenTracing span context or {@code null} if no span is currently active.
+     *            An implementation should use this as the parent for any span it creates for tracing
+     *            the execution of this operation.
+     * @return A future indicating whether the operation succeeded or not.
+     * @throws NullPointerException if device id or adapter instance id is {@code null}.
+     */
+    Future<Void> registerCommandConsumer(String tenantId, String deviceId, String adapterInstanceId, Duration lifespan,
+            SpanContext context);
+
+    /**
+     * Unregisters a command consumer for a device.
+     * <p>
+     * The registration entry is only deleted if its value contains the given protocol adapter instance id.
+     *
+     * @param tenantId The tenant id.
+     * @param deviceId The device id.
+     * @param adapterInstanceId The protocol adapter instance id that the entry to be removed has to contain.
+     * @param context The currently active OpenTracing span context or {@code null} if no span is currently active.
+     *            An implementation should use this as the parent for any span it creates for tracing
+     *            the execution of this operation.
+     * @return A future indicating the outcome of the operation, with its value indicating whether the protocol
+     *         adapter instance value was removed or not.
+     *         <p>
+     *         The future will be failed with a {@link org.eclipse.hono.client.ServiceInvocationException} if there
```

Comment:
The future will be failed with a {@code org.eclipse.hono.client.ServiceInvocationException} if there